File: services/client/src/client/protocol.go
package client

import (
	"bytes"
	"encoding/binary"
)

const ENDOFBETS_BYTES = 2
const BET_BYTES = 1
const ACK_BYTES = 3
const AGENCYID_BYTES = 1
const FIRSTNAME_BYTES = 2
const LASTNAME_BYTES = 3
const DOCUMENT_BYTES = 4
const BIRTHDATE_BYTES = 5
const NUMBER_BYTES = 6
const BATCH_BYTES = 4
const TYPE_BYTES = 2
const SIZE_BYTES = 2

type Message interface {
	Type() int
}

type ErrResponseMismatch struct{}

func (ErrResponseMismatch) Error() string {
	return "received response does not match expected message type"
}

type Bet struct {
	agencyId  []byte
	firstName []byte
	lastName  []byte
	document  []byte
	birthdate []byte
	number    []byte
	byteBet   []byte
}

func (bet *Bet) Type() int {
	return BET_BYTES
}

type Ack struct{}

func (ack Ack) Type() int {
	return ACK_BYTES
}

type Batch struct {
	bets []*Bet
}

func (Batch) Type() int {
	return BATCH_BYTES
}

func encodeField(fieldType int, value []byte) []byte {
	header := make([]byte, TYPE_BYTES+SIZE_BYTES)
	binary.BigEndian.PutUint16(header[:TYPE_BYTES], uint16(fieldType))
	binary.BigEndian.PutUint16(header[TYPE_BYTES:], uint16(len(value)))
	return append(header, value...)
}

func createBet(data []byte, agencyId string) *Bet {
	splitData := bytes.Split(data, []byte(","))
	firstName := splitData[0]
	lastName := splitData[1]
	document := splitData[2]
	birthdate := splitData[3]
	number := splitData[4]

	value := encodeField(AGENCYID_BYTES, []byte(agencyId))
	value = append(value, encodeField(FIRSTNAME_BYTES, firstName)...)
	value = append(value, encodeField(LASTNAME_BYTES, lastName)...)
	value = append(value, encodeField(DOCUMENT_BYTES, document)...)
	value = append(value, encodeField(BIRTHDATE_BYTES, birthdate)...)
	value = append(value, encodeField(NUMBER_BYTES, number)...)

	byteBet := encodeField(BET_BYTES, value)

	return &Bet{
		agencyId:  []byte(agencyId),
		firstName: firstName,
		lastName:  lastName,
		document:  document,
		birthdate: birthdate,
		number:    number,
		byteBet:   byteBet,
	}
}

func splitMsg(data []byte) (msgType []byte, msg []byte, rest []byte) {
	msgType, size := parseHeader(data[:TYPE_BYTES+SIZE_BYTES])
	totalLen := TYPE_BYTES + SIZE_BYTES + size
	return msgType, data[:totalLen], data[totalLen:]
}

func decodeBet(msg []byte) *Bet {
	value := msg[TYPE_BYTES+SIZE_BYTES:]
	bet := &Bet{byteBet: msg}
	for len(value) > 0 {
		fieldType, fieldMsg, rest := splitMsg(value)
		fieldValue := fieldMsg[TYPE_BYTES+SIZE_BYTES:]
		if fieldType[0] == 0 && fieldType[1] == AGENCYID_BYTES {
			bet.agencyId = fieldValue
		} else if fieldType[0] == 0 && fieldType[1] == FIRSTNAME_BYTES {
			bet.firstName = fieldValue
		} else if fieldType[0] == 0 && fieldType[1] == LASTNAME_BYTES {
			bet.lastName = fieldValue
		} else if fieldType[0] == 0 && fieldType[1] == DOCUMENT_BYTES {
			bet.document = fieldValue
		} else if fieldType[0] == 0 && fieldType[1] == BIRTHDATE_BYTES {
			bet.birthdate = fieldValue
		} else if fieldType[0] == 0 && fieldType[1] == NUMBER_BYTES {
			bet.number = fieldValue
		}
		value = rest
	}
	return bet
}

func decodeBatch(value []byte) *Batch {
	var bets []*Bet
	for len(value) > 0 {
		_, msg, rest := splitMsg(value)
		bets = append(bets, decodeBet(msg))
		value = rest
	}
	return &Batch{bets: bets}
}

func parseHeader(header []byte) (msgType []byte, size int) {
	msgType = header[:TYPE_BYTES]
	size = int(binary.BigEndian.Uint16(header[TYPE_BYTES:]))
	return msgType, size
}

func encodeBatch(bets []*Bet) []byte {
	var value []byte
	for _, bet := range bets {
		value = append(value, bet.byteBet...)
	}
	return encodeField(BATCH_BYTES, value)
}

func parseMessage(header []byte, value []byte) (Message, error) {
	msgType := header[:TYPE_BYTES]

	if msgType[0] == 0 && msgType[1] == BET_BYTES {
		return &Bet{byteBet: append(header, value...)}, nil
	} else if msgType[0] == 0 && msgType[1] == ACK_BYTES {
		return Ack{}, nil
	} else if msgType[0] == 0 && msgType[1] == BATCH_BYTES {
		return decodeBatch(value), nil
	}

	return nil, nil
}

File: services/server/src/protocol/protocol.py
from lottery import Bet

BYTE_ORDER = "big"

TYPE_BYTES = 2
SIZE_BYTES = 2

BET_BYTES = 1
ENDOFBETS_BYTES = 2
ACK_BYTES = 3
BATCH_BYTES = 4

AGENCYID_BYTES = 1
FIRSTNAME_BYTES = 2
LASTNAME_BYTES = 3
DOCUMENT_BYTES = 4
BIRTHDATE_BYTES = 5
NUMBER_BYTES = 6


class EndOfBets:
    def __init__(self, agency_id):
        self.agency_id = agency_id


def parse_header(header):
    msg_type = int.from_bytes(header[:TYPE_BYTES], BYTE_ORDER)
    size = int.from_bytes(header[TYPE_BYTES:], BYTE_ORDER)
    return msg_type, size


def split_msg(data):
    msg_type, size = parse_header(data[: TYPE_BYTES + SIZE_BYTES])
    total_len = TYPE_BYTES + SIZE_BYTES + size
    return msg_type, data[:total_len], data[total_len:]


def decode_bet(value):
    agency_id = None
    first_name = None
    last_name = None
    document = None
    birthdate = None
    number = None

    while len(value) > 0:
        field_type, field_msg, rest = split_msg(value)
        field_value = field_msg[TYPE_BYTES + SIZE_BYTES :]

        if field_type == AGENCYID_BYTES:
            agency_id = field_value
        elif field_type == FIRSTNAME_BYTES:
            first_name = field_value
        elif field_type == LASTNAME_BYTES:
            last_name = field_value
        elif field_type == DOCUMENT_BYTES:
            document = field_value
        elif field_type == BIRTHDATE_BYTES:
            birthdate = field_value
        elif field_type == NUMBER_BYTES:
            number = field_value

        value = rest

    return Bet(
        agency_id=int(agency_id.decode()),
        first_name=first_name.decode(),
        last_name=last_name.decode(),
        document=int(document.decode()),
        birthdate=birthdate.decode(),
        number=int(number.decode()),
    )


def decode_batch(value):
    bets = []
    while len(value) > 0:
        _, msg, rest = split_msg(value)
        bets.append(decode_bet(msg[TYPE_BYTES + SIZE_BYTES :]))
        value = rest
    return bets


def parse_message(msg_type, value):
    if msg_type == BET_BYTES:
        return decode_bet(value)
    elif msg_type == ENDOFBETS_BYTES:
        return EndOfBets(agency_id=int(value.decode()))
    elif msg_type == BATCH_BYTES:
        return decode_batch(value)


def encode_field(field_type, value):
    header = field_type.to_bytes(TYPE_BYTES, BYTE_ORDER) + len(value).to_bytes(SIZE_BYTES, BYTE_ORDER)
    return header + value


def encode_bet(bet):
    value = encode_field(AGENCYID_BYTES, str(bet.agency_id).encode())
    value += encode_field(FIRSTNAME_BYTES, bet.first_name.encode())
    value += encode_field(LASTNAME_BYTES, bet.last_name.encode())
    value += encode_field(DOCUMENT_BYTES, str(bet.document).encode())
    value += encode_field(BIRTHDATE_BYTES, bet.birthdate.encode())
    value += encode_field(NUMBER_BYTES, str(bet.number).encode())
    return encode_field(BET_BYTES, value)


def encode_batch(bets):
    value = b"".join(encode_bet(bet) for bet in bets)
    return encode_field(BATCH_BYTES, value)
